public/app/features/explore/QueryRows.tsx

```diff
@@ -51,18 +51,24 @@ export const QueryRows = ({ exploreId }: Props) => {
 
   const onChange = useCallback(
     async (newQueries: DataQuery[]) => {
-      dispatch(changeQueriesAction({ queries: newQueries, exploreId }));
+      let queriesImported = false;
 
       for (const newQuery of newQueries) {
         for (const oldQuery of queries) {
           if (newQuery.refId === oldQuery.refId && newQuery.datasource?.type !== oldQuery.datasource?.type) {
-            const queryDatasource = await getDataSourceSrv().get(newQuery.datasource);
+            const queryDatasource = await getDataSourceSrv().get(oldQuery.datasource);
```

Comment:
Nice catch! How come it works in `main`? It looks like use incorrect source data source? Or is it broken in `main`? 🤔 Could you add a test for it? Also at first sight it looks like the whole callback could be a redux action.

Reply:
it doesn't work in main :)

the reason it seems to work is because you are probably switching between prom and loki, and because of the description above, the `expr` property is kept when switching between them. 

if you modify `public/app/plugins/datasource/prometheus/language_utils.ts` to return `'something'` on line 283, you should see that `{something}` as the result of the import. in main you won't see that happening.

as per adding tests, as mentioning during the standup i tried to implement one but there are a lot of moving parts and i was having an hard time doing it. i also just moved the logic into a dedicated thunk, which should possibly make testing (and isolating) this easier, but again there are a lot of moving bits (especially around mocking datasourceSrv and datasources with query import and export support). i will spend more time on this later